In [5]:
# SUPPRESSING WARNINGS FOR CLEANER OUTPUT.
import warnings
warnings.filterwarnings("ignore")

In [6]:
# IMPORT PANDAS LIBRARY AND ALIAS IT AS 'pd'
import pandas as pd
# LOAD THE Salary_DATASET FROM THE SPECIFIED FILE PATH
le = pd.read_csv('life_expectancy.csv')
# DISPLAY THE FIRST FEW ROWS OF THE DATASET TO GET AN OVERVIEW
le.head()


Country  Year      Status  Life expectancy   Adult Mortality  \
0  Afghanistan  2015  Developing              65.0            263.0   
1  Afghanistan  2014  Developing              59.9            271.0   
2  Afghanistan  2013  Developing              59.9            268.0   
3  Afghanistan  2012  Developing              59.5            272.0   
4  Afghanistan  2011  Developing              59.2            275.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   ...  \
0             62     0.01               71.279624         65.0      1154  ...   
1             64     0.01               73.523582         62.0       492  ...   
2             66     0.01               73.219243         64.0       430  ...   
3             69     0.01               78.184215         67.0      2787  ...   
4             71     0.01                7.097109         68.0      3013  ...   

   Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  Population  \
0    6.0               8.16         65.0        0.1  584.259210  33736494.0   
1   58.0               8.18         62.0        0.1  612.696514    327582.0   
2   62.0               8.13         64.0        0.1  631.744976  31731688.0   
3   67.0               8.52         67.0        0.1  669.959000   3696958.0   
4   68.0               7.87         68.0        0.1   63.537231   2978599.0   

    thinness  1-19 years   thinness 5-9 years  \
0                   17.2                 17.3   
1                   17.5                 17.5   
2                   17.7                 17.7   
3                   17.9                 18.0   
4                   18.2                 18.2   

   Income composition of resources  Schooling  
0                            0.479       10.1  
1                            0.476       10.0  
2                            0.470        9.9  
3                            0.463        9.8  
4                            0.454        9.5  

[5 rows x 22 columns]

In [7]:
# DISPLAY THE INFORMATION OF THE DATASET
le.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          2938 non-null   object 
 1   Year                             2938 non-null   int64  
 2   Status                           2938 non-null   object 
 3   Life expectancy                  2928 non-null   float64
 4   Adult Mortality                  2928 non-null   float64
 5   infant deaths                    2938 non-null   int64  
 6   Alcohol                          2744 non-null   float64
 7   percentage expenditure           2938 non-null   float64
 8   Hepatitis B                      2385 non-null   float64
 9   Measles                          2938 non-null   int64  
 10   BMI                             2904 non-null   float64
 11  under-five deaths                2938 non-null   int64  
 12  Polio               

In [8]:
rows, columns = le.shape

In [9]:
print("\nThe dataset has", rows, "rows and", columns, "columns.")


The dataset has 2938 rows and 22 columns.


In the above imported Dataset, we have 2938 instances with 22 features.

The features are as follows:

Features:

Country: Nation name.
Year: Year.
Status: Developing/Developed.
Life expectancy: Age in years.
Adult Mortality: Mortality rate.
infant deaths: Infant deaths/1000 pop.
Alcohol: Litres of alcohol consumption.
percentage expenditure: % of GDP.
Hepatitis B: HepB immunization (%).
Measles: Reported cases.
BMI: Average.
under-five deaths: Deaths / 1000 pop.
Polio: Polio immunization (%).
Total expenditure: Gov health expenditure (%).
Diphtheria: DTP3 immunization (%).
HIV/AIDS: Deaths / 1000 births.
GDP: USD.
Population: Number of citizens.
thinness 1-19 years: Prevalence (%).
thinness 5-9 years: Prevalence (%).
Income composition of resources: HDI index.
Schooling: Years.

The dataset includes float64 (16), int64 (4), and object (2) data types, and has missing values.

The dataset comprises a mix of data types: float64 (16 columns), int64 (4 columns), and object (2 columns). There are missing values in some columns, as indicated by the "Non-Null Count" being less than 2938 for those columns.

In [11]:
# 2. What attributes/features are continuous valued? (Using .dtypes and inspection)

# Get the data types of all columns
data_types = le.dtypes

# List of continuous features (identified by data type float64 or int64, but exclude Year)
continuous_features = data_types[((data_types == 'float64') | (data_types == 'int64')) & (data_types.index != 'Year')].index.tolist()

print("\nContinuous Features:")
for feature in continuous_features:
    print(f"- {feature}")


Continuous Features:
- Life expectancy 
- Adult Mortality
- infant deaths
- Alcohol
- percentage expenditure
- Hepatitis B
- Measles 
-  BMI 
- under-five deaths 
- Polio
- Total expenditure
- Diphtheria 
-  HIV/AIDS
- GDP
- Population
-  thinness  1-19 years
-  thinness 5-9 years
- Income composition of resources
- Schooling


In [12]:
# 3. Which attributes are categorical? (Based on data types)
categorical_features = data_types[data_types == 'object'].index.tolist()
print("\nCategorical Features:")
for feature in categorical_features:
    print(f"- {feature}")



Categorical Features:
- Country
- Status


### Question B:
#### Visualization and Summary Statistics
- **Q.1** Visualization and Summary Statistics - 3 pts  
- **Q.2** Special Treatment Needed - 2 pts

In [14]:
# DISPLAY SUMMARY STATISTICS OF THE DATAFRAME FOR EACH ATTRIBUTE
le.describe()

Year  Life expectancy   Adult Mortality  infant deaths  \
count  2938.000000       2928.000000      2928.000000    2938.000000   
mean   2007.518720         69.224932       164.796448      30.303948   
std       4.613841          9.523867       124.292079     117.926501   
min    2000.000000         36.300000         1.000000       0.000000   
25%    2004.000000         63.100000        74.000000       0.000000   
50%    2008.000000         72.100000       144.000000       3.000000   
75%    2012.000000         75.700000       228.000000      22.000000   
max    2015.000000         89.000000       723.000000    1800.000000   

           Alcohol  percentage expenditure  Hepatitis B       Measles   \
count  2744.000000             2938.000000  2385.000000    2938.000000   
mean      4.602861              738.251295    80.940461    2419.592240   
std       4.052413             1987.914858    25.070016   11467.272489   
min       0.010000                0.000000     1.000000       0.000000   
25%       0.877500                4.685343    77.000000       0.000000   
50%       3.755000               64.912906    92.000000      17.000000   
75%       7.702500              441.534144    97.000000     360.250000   
max      17.870000            19479.911610    99.000000  212183.000000   

              BMI   under-five deaths         Polio  Total expenditure  \
count  2904.000000         2938.000000  2919.000000         2712.00000   
mean     38.321247           42.035739    82.550188            5.93819   
std      20.044034          160.445548    23.428046            2.49832   
min       1.000000            0.000000     3.000000            0.37000   
25%      19.300000            0.000000    78.000000            4.26000   
50%      43.500000            4.000000    93.000000            5.75500   
75%      56.200000           28.000000    97.000000            7.49250   
max      87.300000         2500.000000    99.000000           17.60000   

       Diphtheria      HIV/AIDS            GDP    Population  \
count  2919.000000  2938.000000    2490.000000  2.286000e+03   
mean     82.324084     1.742103    7483.158469  1.275338e+07   
std      23.716912     5.077785   14270.169342  6.101210e+07   
min       2.000000     0.100000       1.681350  3.400000e+01   
25%      78.000000     0.100000     463.935626  1.957932e+05   
50%      93.000000     0.100000    1766.947595  1.386542e+06   
75%      97.000000     0.800000    5910.806335  7.420359e+06   
max      99.000000    50.600000  119172.741800  1.293859e+09   

        thinness  1-19 years   thinness 5-9 years  \
count            2904.000000          2904.000000   
mean                4.839704             4.870317   
std                 4.420195             4.508882   
min                 0.100000             0.100000   
25%                 1.600000             1.500000   
50%                 3.300000             3.300000   
75%                 7.200000             7.200000   
max                27.700000            28.600000   

       Income composition of resources    Schooling  
count                      2771.000000  2775.000000  
mean                          0.627551    11.992793  
std                           0.210904     3.358920  
min                           0.000000     0.000000  
25%                           0.493000    10.100000  
50%                           0.677000    12.300000  
75%                           0.779000    14.300000  
max                           0.948000    20.700000

In [15]:
print("\nMissing Values in Each Column:")
print(le.isnull().sum())


Missing Values in Each Column:
Country                              0
Year                                 0
Status                               0
Life expectancy                     10
Adult Mortality                     10
infant deaths                        0
Alcohol                            194
percentage expenditure               0
Hepatitis B                        553
Measles                              0
 BMI                                34
under-five deaths                    0
Polio                               19
Total expenditure                  226
Diphtheria                          19
 HIV/AIDS                            0
GDP                                448
Population                         652
 thinness  1-19 years               34
 thinness 5-9 years                 34
Income composition of resources    167
Schooling                          163
dtype: int64


In [42]:
import numpy as np
for col in le.select_dtypes(include=np.number):
    le[col] = le[col].fillna(le[col].mean())

In [50]:
le.isna().sum()
#null values eliminated

Country                         0
Year                            0
Status                          0
Lifeexpectancy                  0
AdultMortality                  0
infantdeaths                    0
Alcohol                         0
percentageexpenditure           0
HepatitisB                      0
Measles                         0
BMI                             0
underfivedeaths                 0
Polio                           0
Totalexpenditure                0
Diphtheria                      0
HIVAIDS                         0
GDP                             0
Population                      0
thinness119years                0
thinness59years                 0
Incomecompositionofresources    0
Schooling                       0
dtype: int64

In [52]:
# Rename Columns (To remove spaces and special characters in column names)
le.columns = le.columns.str.replace('[^A-Za-z0-9_]+', '', regex=True)
le.rename(columns={'Lifeexpectancy': 'Lifeexpectancy',
                   'Incomecompositionofresources': 'Incomecompositionofresources',
                   'thinness119years': 'thinness119years',
                   'thinness59years': 'thinness59years'}, inplace=True)

In [56]:
le.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Country                       2938 non-null   object 
 1   Year                          2938 non-null   int64  
 2   Status                        2938 non-null   object 
 3   Lifeexpectancy                2938 non-null   float64
 4   AdultMortality                2938 non-null   float64
 5   infantdeaths                  2938 non-null   int64  
 6   Alcohol                       2938 non-null   float64
 7   percentageexpenditure         2938 non-null   float64
 8   HepatitisB                    2938 non-null   float64
 9   Measles                       2938 non-null   int64  
 10  BMI                           2938 non-null   float64
 11  underfivedeaths               2938 non-null   int64  
 12  Polio                         2938 non-null   float64
 13  Tot

In [60]:
corr_le=le.drop(['Country','Status'],axis=1)

In [64]:
corr_le.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 20 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Year                          2938 non-null   int64  
 1   Lifeexpectancy                2938 non-null   float64
 2   AdultMortality                2938 non-null   float64
 3   infantdeaths                  2938 non-null   int64  
 4   Alcohol                       2938 non-null   float64
 5   percentageexpenditure         2938 non-null   float64
 6   HepatitisB                    2938 non-null   float64
 7   Measles                       2938 non-null   int64  
 8   BMI                           2938 non-null   float64
 9   underfivedeaths               2938 non-null   int64  
 10  Polio                         2938 non-null   float64
 11  Totalexpenditure              2938 non-null   float64
 12  Diphtheria                    2938 non-null   float64
 13  HIV

In [ ]:
# Special Treatment of Attributes
# Apply log transformation to reduce skewness
le_cleaned['Population'] = np.log1p(le_cleaned['Population'])
le_cleaned['GDP'] = np.log1p(le_cleaned['GDP'])


In [ ]:
# Create a new column to flag missing 'Alcohol' values
le_cleaned['Alcohol_Missing'] = le_cleaned['Alcohol'].isnull().astype(int)

In [ ]:
# Fill missing 'Alcohol' values with a specific value (e.g., mean)
le_cleaned['Alcohol'].fillna(le_cleaned['Alcohol'].mean(), inplace=True)


In [ ]:
# Winsorize 'Lifeexpectancy' to handle outliers
q1 = le_cleaned['Lifeexpectancy'].quantile(0.05)
q3 = le_cleaned['Lifeexpectancy'].quantile(0.95)
le_cleaned['Lifeexpectancy'] = le_cleaned['Lifeexpectancy'].clip(q1, q3)


In [ ]:
# Display the cleaned data
print("\nCleaned Data:")
print(le_cleaned.head())



## [Data Preprocessing and Special Treatments](pplx://action/followup)

The life expectancy dataset was preprocessed to address missing values, skewness, and outliers, aiming to improve the quality and suitability of the data for subsequent analysis and modeling.

**[1. Handling Missing Data](pplx://action/followup):**

   *   Rows with more than three missing values were removed to ensure data completeness and reliability.
   *   Remaining missing values in numerical columns were imputed using the mean of each respective column. A missing indicator column, `Alcohol_Missing`, was created to flag missing entries in the `Alcohol` column before imputation. This aims to capture potential information related to the missingness itself.

**[2. Addressing Skewness](pplx://action/followup):**

   *   The `Population` and `GDP` variables exhibited significant right skewness. To mitigate this, a log transformation was applied using the `np.log1p()` function.  This transformation helps normalize the distributions, making them more suitable for linear models and reducing the influence of extreme values.

**[3. Outlier Management](pplx://action/followup):**

   *   Outliers in the `Life expectancy` column were addressed using winsorization. Values below the 5th percentile and above the 95th percentile were capped at those respective percentile values. This approach reduces the impact of extreme values without completely removing them from the dataset.

### Q.3 Analyze the Relationships:
- **Objective**: Analyze the relationships between the data attributes, and between the data attributes and the label.
- **Steps**:
  1. Compute the **Pearson Correlation Coefficient (PCC)** to measure the linear correlation between the attributes.
  2. Generate **scatter plots** to visualize the relationships between the attributes and the label.

This process helps identify potential correlations, trends, or patterns in the data that can be used for further analysis or modeling.


In [ ]:
#1.Calculate the Pearson Correlation Coefficient (PCC) Matrix:

In [ ]:
import seaborn as sns

In [ ]:
# Calculate the correlation matrix
correlation_matrix = le_cleaned.corr(numeric_only=True)

In [ ]:
# Display the correlation matrix using a heatmap
plt.figure(figsize=(16, 12))
sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', fmt=".2f") #fmt for formatting of values
plt.title("Pearson Correlation Coefficient (PCC) Matrix")
plt.show()

In [ ]:
#2 Example: Scatter plots for 'Schooling' and 'GDP' vs. 'Lifeexpectancy'

#Scatter plot: Schooling vs. Life Expectancy
plt.figure(figsize=(8, 6))
sns.scatterplot(x='Schooling', y='Lifeexpectancy', data=le_cleaned)
plt.title("Scatter Plot of Schooling vs. Life Expectancy")
plt.xlabel("Schooling (Years)")
plt.ylabel("Life Expectancy (Years)")
plt.grid(True)
plt.show()

# Scatter plot: GDP vs. Life Expectancy
plt.figure(figsize=(8, 6))
sns.scatterplot(x='GDP', y='Lifeexpectancy', data=le_cleaned)
plt.title("Scatter Plot of GDP vs. Life Expectancy")
plt.xlabel("GDP (Log Transformed)")
plt.ylabel("Life Expectancy (Years)")
plt.grid(True)
plt.show()


### Q4: Select 20% of the Data for Testing
- **Objective**: Select 20% of the data for testing.
- **Steps**:
  1. Split the dataset into two portions: 80% for training and 20% for testing.
  2. **Description**: Explain the method used to select the test data, e.g., random sampling, stratified sampling, etc.
  3. **Verification**: Verify that the test portion of the data is representative of the entire dataset. This can be done by comparing summary statistics, distributions, or visualizing both training and testing sets to check for any significant differences.


In [ ]:
# IMPORT NECESSARY LIBRARY FOR TRAIN-TEST SPLIT
from sklearn.model_selection import train_test_split